### DISCUSSION TITLE
Error "SSL_write: Connection reset by peer" after receiving HTTP 200

### DISCUSSION THREAD

**Discussion by ribtoks** [curl/curl#8117] [Q&A] [answered]
Hi

I'm sending a file with libcurl to Backblaze/S3 via presigned URL. After sending all the data, server replies `200`, but then the transfer fails with an error `OpenSSL SSL_write: Connection reset by peer, errno 104`.

Here's the full log:

```
... tls handshake ...

16:51:39.162 info T#1074707 Connectivity::my_trace - "== Info: SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256"
16:51:39.162 info T#1074707 Connectivity::my_trace - "== Info: ALPN, server did not agree to a protocol"

.... millions of logs of sending 5 and 1 byte ....

15:51:01.528 debug T#1058820 Connectivity::dump - "=> Send SSL data, 0000000005 bytes (0x00000005)\n...@."
15:51:01.528 debug T#1058820 Connectivity::dump - "=> Send SSL data, 0000000001 bytes (0x00000001)\n."
15:51:01.646 debug T#1058820 Connectivity::dump - "<= Recv SSL data, 0000000005 bytes (0x00000005)\n..../"
15:51:01.646 debug T#1058820 Connectivity::dump - "<= Recv SSL data, 0000000001 bytes (0x00000001)\n."
15:51:01.646 info T#1058820 Connectivity::my_trace - "== Info: Mark bundle as not supporting multiuse"
15:51:01.646 debug T#1058820 Connectivity::dump - "<= Recv header, 0000000015 bytes (0x0000000f)\nHTTP/1.1 200 .."
15:51:01.646 debug T#1058820 Connectivity::dump - "<= Recv header, 0000000036 bytes (0x00000024)\nx-amz-request-id: xyz.."
15:51:01.646 debug T#1058820 Connectivity::dump - "<= Recv header, 0000000047 bytes (0x0000002f)\nx-amz-id-2: abc.."
15:51:01.646 debug T#1058820 Connectivity::dump - "<= Recv header, 0000000042 bytes (0x0000002a)\nETag: \"abcxyz\".."
15:51:01.646 debug T#1058820 Connectivity::dump - "<= Recv header, 0000000046 bytes (0x0000002e)\nCache-Control: max-age=0, no-cache, no-store.."
15:51:01.647 debug T#1058820 Connectivity::dump - "<= Recv header, 0000000026 bytes (0x0000001a)\nContent-Type: text/plain.."
15:51:01.647 debug T#1058820 Connectivity::dump - "<= Recv header, 0000000019 bytes (0x00000013)\nContent-Length: 0.."
15:51:01.647 debug T#1058820 Connectivity::dump - "<= Recv header, 0000000037 bytes (0x00000025)\nDate: Wed, 08 Dec 2021 13:51:01 GMT.."
15:51:01.647 debug T#1058820 Connectivity::dump - "<= Recv header, 0000000002 bytes (0x00000002)\n.."
15:51:01.647 debug T#1058820 Connectivity::dump - "=> Send SSL data, 0000000005 bytes (0x00000005)\n...@."
15:51:01.647 debug T#1058820 Connectivity::dump - "=> Send SSL data, 0000000001 bytes (0x00000001)\n."
15:51:01.647 info T#1058820 Connectivity::my_trace - "== Info: OpenSSL SSL_write: Connection reset by peer, errno 104"
15:51:01.647 info T#1058820 Connectivity::my_trace - "== Info: Closing connection 0"
15:51:01.647 warning T#1058820 {} - curl_easy_perform() failed with code 200 : "Failed sending data to the peer" [rpt_me]
```

Sending code uses:
- headers `Content-Length`, `Content-Type` and `Expect: 100-continue` (result is the same if I use or not use `Content-Length`)
- options `CURLOPT_UPLOAD`, `CURLOPT_INFILESIZE_LARGE`, `CURLOPT_READDATA` and `CURLOPT_READFUNCTION`
- a chunk of the file is sent using `fopen()` and `fseek()` and then chunk size specified in `INFILESIZE`

(I can provide the full sending code if needed, just omitted it for brevity since it's more or less default from libcurl examples)

From the logs it seems that we want to send more, but server "had enough" and it closed the connection.

### Details

OS: Ubuntu 20.04
libcurl: 7.80
OpenSSL: 1.1.1l

### Question

My question is if this is me who is misusing curl to do the transfer or the error is actually on the provider side? (also miniscule chance that there's some issue in libcurl, but I wouldn't be sure about that).

Thank you very much for your help.

**Comment by jay** [comment, marked as answer]
If the server says 200 ok I'd assume it's received what was expected. Servers sometimes reply prematurely on error 4xx but I don't recall seeing one reply 200 prematurely. I'd change your read callback to keep track of the total amount being sent and see if it exceeds content length.

>  (result is the same if I use or not use Content-Length)

we'd have to see the code for this, preferably a self contained example


edit: i mean _prematurely_ from our point of view, maybe not the server's.

**Reply by ribtoks** [reply]
@jay Thanks for the idea. I can confirm that the problem is resolved and the issue was that `read_callback()` was returning more data than specified in the `Content-Length`. The reason it happened was that I incorrectly assumed that curl will read only until `INFILESIZE` bytes using the `read_callback()`.

**Comment by bagder** [comment]
errno 104 is `ECONNRESET` the server seems to unexpectedly reset the connection

**Reply by ribtoks** [reply]
Thank you @bagder for replying. The problem was that I did not return `0` from `read_callback()` at the right time and it caused libcurl sending more data after server closed the connection.
